### PR TITLE
chore: import a small(ish) DAO app

### PR DESCRIPTION
**Pull request by ggreif**
This imports @chenyan-dfinity's [simple `dao` dApp](https://github.com/dfinity/examples/tree/master/motoko/basic_dao/src), adding a few tweaks to make it palatable for the `test/perf` framework
- files renamed and partially moved to `dao` directory
- hard-coded `init` parameters
- hard-coded the timestamps
- example proposal filling by query.

This PR also imports/updates a few files in `test/perf/assetstorage` to be in-sync with `motoko-base`.

------
Obtaining the differences
```
$ curl https://raw.githubusercontent.com/dfinity/examples/master/motoko/basic_dao/src/Main.mo | diff -u - test/perf/dao.mo
$ curl https://raw.githubusercontent.com/dfinity/examples/master/motoko/basic_dao/src/Types.mo | diff -u - test/perf/dao/dao-types.mo
```

### CODE REVIEW

File: test/perf/ok/dao.drun-run.ok [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ingress Completed: Reply: 0x4449444c016b029cc2017de58eb402710100012c43616c6c6572206e6565647320616e206163636f756e7420746f207375626d697420612070726f706f73616c
+Ok: Reply: 0x4449444c086d016c08dbb7017db793e6d10307abfec2a4060591ecada00804d6a9bbae0a7cb4bfd4960b68f18d9dc10b078effd6e90e026c03e1edeb4a71b3c4b1f20468c7ebc4d009036d7b6b06ca87d0cd047ffefe8fdc0a7fb2d5bc880b7fe7fadec00c7fbd8bfbbf0e71e1aca5d90f7f6e066c02006801056c01b9ef9380087d010000
+Ok: Reply: 0x4449444c036d016c02b3b0dac30368daae9c8f09026c01b9ef9380087d010000
+Ok: Reply: 0x4449444c086e016c08dbb7017db793e6d10307abfec2a4060591ecada00804d6a9bbae0a7cb4bfd4960b68f18d9dc10b078effd6e90e026c03e1edeb4a71b3c4b1f20468c7ebc4d009036d7b6b06ca87d0cd047ffefe8fdc0a7fb2d5bc880b7fe7fadec00c7fbd8bfbbf0e71e1aca5d90f7f6e066c02006801056c01b9ef9380087d010000
+Ok: Reply: 0x4449444c086e016c08dbb7017db793e6d10307abfec2a4060591ecada00804d6a9bbae0a7cb4bfd4960b68f18d9dc10b078effd6e90e026c03e1edeb4a71b3c4b1f20468c7ebc4d009036d7b6b06ca87d0cd047ffefe8fdc0a7fb2d5bc880b7fe7fadec00c7fbd8bfbbf0e71e1aca5d90f7f6e066c02006801056c01b9ef9380087d010000
```
Comment (ggreif) [inline test/perf/ok/dao.drun-run.ok:12]: this is the response to `get_proposal 1`, but seems to be identical to above. @chenyan-dfinity this is suspicious

File: test/perf/ok/dao.drun-run.ok [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+Ok: Reply: 0x4449444c086e016c08dbb7017db793e6d10307abfec2a4060591ecada00804d6a9bbae0a7cb4bfd4960b68f18d9dc10b078effd6e90e026c03e1edeb4a71b3c4b1f20468c7ebc4d009036d7b6b06ca87d0cd047ffefe8fdc0a7fb2d5bc880b7fe7fadec00c7fbd8bfbbf0e71e1aca5d90f7f6e066c02006801056c01b9ef9380087d010000
```
Comment (ggreif) [inline test/perf/ok/dao.drun-run.ok:13]: Ditto here `get_proposal 15` shouldn't exist. I'd expect a `null` response.

File: test/perf/ok/dao.drun-run.ok [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+Ok: Reply: 0x4449444c026c02ea99cff204018882e69c0a7d6b02c1c0017fa7d2f0027f01000101
+ingress Completed: Reply: 0x4449444c026b029cc20101e58eb402716b06ca87d0cd047ffefe8fdc0a7fb2d5bc880b7fe7fadec00c7fbd8bfbbf0e71e1aca5d90f7f0100011c4e6f2070726f706f73616c2077697468204944203120657869737473
+Ok: Reply: 0x4449444c086e016c08dbb7017db793e6d10307abfec2a4060591ecada00804d6a9bbae0a7cb4bfd4960b68f18d9dc10b078effd6e90e026c03e1edeb4a71b3c4b1f20468c7ebc4d009036d7b6b06ca87d0cd047ffefe8fdc0a7fb2d5bc880b7fe7fadec00c7fbd8bfbbf0e71e1aca5d90f7f6e066c02006801056c01b9ef9380087d010000
```
Comment (ggreif) [inline test/perf/ok/dao.drun-run.ok:16]: Again this is exactly the same answer. Shouldn't it have updated votes for proposal 1?

File: test/perf/ok/dao.drun-run.ok (hunk 1 above)
Comment (ggreif) [reply test/perf/ok/dao.drun-run.ok:12]: Aaah, those are big-typed `null`s. I am to fix this.

**Comment by ggreif** [comment]
Applying the following diff makes the voting process less opaque:
``` diff
$ git diff
diff --git a/test/perf/dao.mo b/test/perf/dao.mo
index e22446190..3528be884 100644
--- a/test/perf/dao.mo
+++ b/test/perf/dao.mo
@@ -8,6 +8,7 @@ import Error "assetstorage/Error";
 import ICRaw "assetstorage/ExperimentalInternetComputer";
 import List "assetstorage/List";
 import Time "assetstorage/Time";
+import {debugPrint} "mo:prim";
 import Types "dao/dao-types";
 
 shared({caller = creator}) actor class DAO() = Self {
@@ -98,6 +99,7 @@ shared({caller = creator}) actor class DAO() = Self {
 
     /// Return the proposal with the given ID, if one exists
     public query func get_proposal(proposal_id: Nat) : async ?Types.Proposal {
+        debugPrint (debug_show (next_proposal_id, proposal_id, proposal_get proposal_id));
         proposal_get(proposal_id)
     };
 
```

**Review by chenyan-dfinity — APPROVED** [review]
Related to this, are we still missing a generic `Int.hash` function in base?

**Pull request merged**